lib/evse_security.cpp

```diff
@@ -569,22 +636,22 @@ GetKeyPairResult EvseSecurity::get_key_pair(LeafCertificateType certificate_type
         key_file = private_key_path;
 
         // We are searching for the full leaf bundle, containing both the leaf and the cso1/2
-        X509CertificateDirectory leaf_directory(cert_dir);
-        const X509CertificateBundle* leaf_fullchain = nullptr;
+        X509CertificateBundle leaf_directory(cert_dir, EncodingFormat::PEM);
+        const std::vector<X509Wrapper>* leaf_fullchain = nullptr;
 
         // Collect the correct bundle
-        leaf_directory.for_each([&certificate, &leaf_fullchain](const X509CertificateBundle& bundle) {
+        leaf_directory.for_each_chain([&](const std::filesystem::path& path, const std::vector<X509Wrapper>& chain) {
             // If we contain the latest valid, we found our generated bundle
-            if (bundle.is_bundle() && bundle.contains_certificate(certificate)) {
-                leaf_fullchain = &bundle;
+            if (chain.size() > 1 && std::find(chain.begin(), chain.end(), certificate) != chain.end()) {
+                leaf_fullchain = &chain;
```

Comment:
Not a comment but a question: Does this mean that we only identify the chain file (containing the leaf and sub-CAs) by the fact that it has multiple certificates, and one of those is the newest certificate in the bundle?

As far as I can tell, the `X509CertificateBundle` constructor would also parse this chain file and add the sub-CAs to the bundle, right? Is it possible that a sub-CA certificate is newer than the leaf certificate it signed?

Reply:
For the first question, yes, the update_leaft_certificate returns a certificate chain that contains leaf->subCa2->subCa2, so it should be the chain which contains the latest valid certificate. If not chain is found, then there's a fallback to the single leaf.

On some systems the V2G root might contain the SubCa's but on others it might not, so we have to detect.

For the second question yes, the bundle can now parse in a directory all files and chains.

For the third question, I don't think that can be the case, since on an update (update_leaf_certificate), we will receive the most recent SubCas packed in the chain.